Containerized integration test environment

ahmetalpbalkan opened the pull request:
Built a containerized environment that looks like an actual VM
and wrote a fake agent which you can invoke the extension binary
through.

Refer to the integration-test/env/README.md for more info.

This helps us test:
- agent binary can handle the directory structure presented by waagent
- agent binary can handle the HandlerEnvironment, log dirs etc
- agent binary decrypts/parses the settings correctly
- agent binary can write status to the right place

By containerizing there will be no residue left on the dev machine
while testing the changes.

paulmey commented on integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv:
```diff
@@ -0,0 +1,33 @@
+    Microsoft CSP Name: Microsoft Enhanced Cryptographic Provider v1.0
+Key Attributes
+    X509v3 Key Usage: 10
+-----BEGIN PRIVATE KEY-----
```
Not really recommended to check in private keys, even if they are for tests. Consider writing a script that generates a keypair and encodes the relevant data for the tests.

paulmey replied:
Generate:
```
openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes -batch
export FP=$(openssl x509 -in cert.pem -fingerprint -noout|sed  's/.*=\|://g')
mv cert.pem $FP.crt
mv key.pem $FP.prv
```
Encode: (put stdout in protected settings)
```
openssl smime -inkey $FP.prv -encrypt -outform DER $FP.crt <script |base64 -w0
```

ahmetalpbalkan replied:
damn you just made it look easy. I guess I will make an utility for this.

paulmey replied:
👍

paulmey commented on integration-test/env/Extension/bin/_placeholder:
```diff
@@ -0,0 +1 @@
+placeholder for git to preserve this directory
```
Seems unnecessary? the COPY command in the docker file creates this directory if it doesn't exist...

ahmetalpbalkan replied:
@paulmey but the directory does not appear in .git at all, for documentation purposes (see README.md) I thought this dir better be present when the repo is checked out?

paulmey replied:
Makes sense, unnecessary for docker, but necessary for the humans... 😉

paulmey commented on test.Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+# Copy the handler files
+COPY HandlerManifest.json ./Extension/
+COPY bin/custom-script-extension ./Extension/bin/
```
I assume that bin/custom-script-extension is created by #29

ahmetalpbalkan replied:
True.

paulmey commented:
I would not check in any private key. LGTM otherwise

ahmetalpbalkan commented:
@paulmey addressed CR comments, feel free to take a look again. Thanks for the openssl tips. I removed the checked-in certs.

I wrote the actual integration tests (it will be another PR) using the [bats](https://github.com/sstephenson/bats) framework and wrote a bunch of  I bash functions to deal with containerized tests and pushing certs/settings to the test container.

paulmey commented:
Awesome! LGTM
Thanks, @ahmetalpbalkan!

ahmetalpbalkan commented:
just added another commit

```
commit 9b7043a120c3dd99bb7a922b8c7b846ca9d90587

    - removed placeholder dirs (create them in dockerfile)
    - change handlermanifest to use the shim
```

ahmetalpbalkan commented:
merging. @boumenot feel free to PTAL sometime
